### PR TITLE
feat(ai): add Anthropic provider with chat parity (1/5)

### PR DESCRIPTION
**Pull request by gariasf**
## Summary

Introduces `Provider::Anthropic` as a first-class LLM provider alongside `Provider::Openai`, implementing the `LlmConcept#chat_response` contract over the official `anthropic` Ruby SDK. **Chat (text + streaming + tool use) is at parity.** Batch ops, PDF, and RAG land in stacked follow-up PRs.

This is **PR 1 of 5** in the Anthropic integration series. Each PR ships independently green.

### Why Anthropic

The Sure assistant currently only speaks to OpenAI. Anthropic offers (a) native PDF understanding without rasterization, (b) explicit prompt-cache control, (c) competitive Sonnet pricing for chat and Haiku for batch ops, and (d) an alternative for users who already standardize on Claude. Adding a parallel provider gives operators a choice without locking anyone in.

### What's in this PR

**New code**

- `Provider::Anthropic` — uses Messages API for sync + streaming responses, wraps SDK errors as `Provider::Anthropic::Error`, records LLM usage with provider `"anthropic"`, threads Langfuse traces with the same span name (`chat_response`) so dashboards stay comparable across providers.
- `Provider::Anthropic::ChatConfig` — builds Messages API requests; tags the system prompt and the final tool definition with `cache_control: { type: "ephemeral" }` (Anthropic's prompt caching).
- `Provider::Anthropic::MessageFormatter` — translates `Chat#conversation_messages` into Anthropic's content-block schema. Each historical assistant turn that used tools is rebuilt as the required `tool_use` block + paired `user`-role `tool_result` block before the next assistant turn. In-flight `function_results` (the responder's second-leg call) get the same treatment.
- `Provider::Anthropic::ChatParser` — converts the SDK's `Message` (or a hash-shaped equivalent) into the shared `Provider::LlmConcept::ChatResponse` Data struct, surfacing both text blocks and `tool_use` blocks as function requests.

**Wiring**

- `Provider::Registry` — new `anthropic` factory; reads `ANTHROPIC_ACCESS_TOKEN` / `ANTHROPIC_API_KEY` / `Setting.anthropic_access_token`, plus optional `ANTHROPIC_BASE_URL` (for Bedrock / Vertex proxies) and `ANTHROPIC_MODEL`. Listed under both `:llm` and the default available_providers set so existing `supports_model?` routing picks the right provider for `claude-*` models.
- `Setting` — adds `anthropic_access_token`, `anthropic_model`, `anthropic_base_url`, and `llm_provider` (default `"openai"`) fields.
- `User#ai_available?` — now returns true when *either* OpenAI or Anthropic is configured; exposes `openai_configured?` / `anthropic_configured?` for consumers.
- `Chat.default_model` — resolves from `Setting.llm_provider`, so installs that flip the provider don't have to update every chat default.
- `Assistant::Responder` — passes a new `conversation_history:` kwarg (raw `Array<Message>`) alongside the existing `messages:` (OpenAI-shaped) so providers without hosted conversation state can rebuild context. The OpenAI provider accepts and ignores the new kwarg — no behavior change.
- `Provider::LlmConcept` — signature extended to document the new kwarg.

### What's *not* in this PR (intentional)

- ❌ `auto_categorize`, `auto_detect_merchants`, `enhance_provider_merchants` — raise a clear `not yet implemented` error on this provider. **PR 2/5** lands these via forced tool calls (`tool_choice: {type: "tool", name: ...}`) plus the cost-ledger migration.
- ❌ `process_pdf`, `extract_bank_statement` — same story. **PR 3/5** uses native `document` content blocks.
- ❌ `Vector::Store::*` / RAG — **PR 4/5** ships a pgvector-backed store so RAG works without Anthropic needing a hosted vector store.
- ❌ Settings UI panel — **PR 5/5** mirrors `_openai_settings.html.erb` and adds the data-retention disclosure (parity with OpenAI's panel).
- ❌ Live VCR cassettes for `chat_response`. Cassettes need to be recorded with a real `ANTHROPIC_API_KEY`; mocked tests cover the wiring in the meantime.

### Configuration (new env vars)

| Env | Setting field | Default |
|---|---|---|
| `ANTHROPIC_ACCESS_TOKEN` *(or)* `ANTHROPIC_API_KEY` | `Setting.anthropic_access_token` | nil |
| `ANTHROPIC_MODEL` | `Setting.anthropic_model` | `claude-sonnet-4-6` |
| `ANTHROPIC_BASE_URL` | `Setting.anthropic_base_url` | nil (uses api.anthropic.com) |
| `ANTHROPIC_REQUEST_TIMEOUT` | — | 600 |
| `ANTHROPIC_MAX_TOKENS` | — | 4096 |
| `LLM_PROVIDER` | `Setting.llm_provider` | `openai` |

### Test plan

- [x] `bin/rails test test/models/provider/anthropic_test.rb` — 8 tests, init / supports_model / supports_pdf_processing / effective_model / chat_response success + error + tool_use paths
- [x] `bin/rails test test/models/provider/anthropic/` — 14 tests across ChatConfig, ChatParser, MessageFormatter
- [x] `bin/rails test test/models/provider/registry_test.rb` — 8 tests, includes new `:anthropic` factory + env-empty fallback to Setting
- [x] `bin/rails test test/models/provider/openai_test.rb` — 27 tests, unchanged
- [x] Full suite: 4358 runs, 18008 assertions, **0 failures**, 26 pre-existing skips, 1 pre-existing env error (libvips unavailable on the dev host)
- [x] `bin/rubocop` clean on all touched files
- [x] `bin/brakeman --no-pager` — 0 security warnings
- [ ] Record live VCR cassettes (follow-up — needs `ANTHROPIC_API_KEY`)
- [ ] System test exercising a chat turn through `Provider::Anthropic` (follow-up — bundled with cassettes)

### Compliance notes

- Anthropic's standard tier does not train on API inputs; 30-day Trust & Safety retention. The settings UI in PR 5/5 surfaces this alongside OpenAI's equivalent disclosure.
- The financial-advice disclaimer already shown in the chat applies equally to either provider.
- `anthropic` gem is MIT-licensed, compatible with Sure's AGPLv3.
- OpenAI integration is untouched.

## Summary by CodeRabbit

* **New Features**
  * Anthropic added as an alternative LLM provider with configurable credentials and optional custom endpoint.

* **Behavior Changes**
  * Default LLM selection now respects the deploy-time provider setting and prefers the configured provider.
  * Chat handling now includes prior conversation history to improve responses.
  * User AI availability now reflects either OpenAI or Anthropic being configured.

* **Chores**
  * Added Anthropic client dependency to the project.

* **Tests**
  * Expanded Anthropic-related test coverage (configuration, parsing, formatting, streaming).

[Review Change Stack](https://app.coderabbit.ai/change-stack/we-promise/sure/pull/1983?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by jjmata** [comment]
Solid foundation PR. A few observations:

**Message format handling:** The `MessageFormatter` rebuilding tool-use turns into the required `tool_use` block + paired `user`-role `tool_result` block is the trickiest part of Anthropic's Messages API (Anthropic requires tool results to be in the `user` role, not `assistant`). Glad to see it's explicitly handled rather than bolted on later.

**Prompt caching:** Attaching `cache_control: { type: "ephemeral" }` to the system prompt and final tool definition in `ChatConfig` is the correct Anthropic caching pattern. System prompts and tool definitions are the highest-leverage targets since they're large and repeated on every turn.

**`conversation_history:` kwarg on Responder:** Backward-compatible (OpenAI path ignores it). Worth confirming that `chat_message_records` is correctly serialized into Anthropic's content-block format by `MessageFormatter` — specifically that tool-use turns from prior assistant messages round-trip correctly when retrieved from the database.

**`Chat.default_model` via `Setting.llm_provider`:** Clean abstraction. One edge case to confirm: if `llm_provider` is `"openai"` but no OpenAI key is configured (while an Anthropic key is), `User#ai_available?` would return true (Anthropic configured), but the chat would try to use an OpenAI model. The PR description says `Chat.default_model` resolves from `Setting.llm_provider`, so if the operator set `llm_provider = "anthropic"`, the default model would be `claude-sonnet-4-6`. That's correct. The edge case only applies if an operator sets `llm_provider = "openai"` with no key, which is a misconfiguration.

**VCR cassettes:** Noted as a follow-up requiring a real `ANTHROPIC_API_KEY`. The mocked tests cover the wiring adequately for a foundation PR.

**Stacked PRs (#1984–#1987):** All currently in draft — appropriate while #1983 awaits review. The series is well-structured with each PR independently green.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by gariasf** [comment]
Thanks for the review, @jjmata. Quick responses on each point:

**Round-trip confirmation** — addressed in `b26306d8`. Added a `ChatParser → ToolCall::Function → MessageFormatter` round-trip test that walks the full path: Anthropic response w/ a `tool_use` block → `ChatFunctionRequest` → `ToolCall::Function.from_function_request` → persisted on the AssistantMessage → `MessageFormatter` rebuild on the next turn. Asserts the original `tool_use.id` survives end-to-end as both the assistant turn's `tool_use.id` and the paired user turn's `tool_result.tool_use_id`. Also added a multi-tool turn test confirming two tool_use blocks render as two `tool_use` blocks followed by two paired `tool_result` blocks (Anthropic's required alternation).

**`Chat.default_model` edge case** — `prefers_anthropic && Provider::Anthropic.configured?` short-circuits the first branch when Anthropic isn't actually configured. The next branch (`elsif Provider::Openai.configured?`) wins. If neither is configured the final `else` returns OpenAI's `effective_model` for backward compat. The reverse case you described — `llm_provider="openai"` + only Anthropic key set — falls through `elsif Provider::Anthropic.configured?` and returns Claude, which is the right call.

**VCR cassettes** — yes, slated as a follow-up bundled with #1985 (PDF tests need a real PDF + key too). Mocked client coverage is the bar for the foundation PR.

Stacked PRs all rebased and pushed earlier today; each independently green.

**Comment by sure-design** [comment]
Three functional blockers still open after the latest commits — flagging for resolution before merge:

**1. Streaming class namespace mismatch (silent failure)**
`stream_chat_response` matches `::Anthropic::Streaming::TextEvent` / `::Anthropic::Streaming::MessageStopEvent`, but `client.messages.stream` yields `Anthropic::Helpers::Streaming::TextEvent` / `Anthropic::Helpers::Streaming::MessageStopEvent`. The `case` branches never fire, so streaming returns empty output silently.

**2. `Provider::Anthropic#chat_response` missing `messages:` kwarg**
`Assistant::Responder` calls `llm.chat_response(..., messages: ..., conversation_history: ...)`. If the Anthropic method signature doesn't accept `messages:`, every Anthropic-backed chat turn raises `ArgumentError: unknown keyword: :messages` at runtime — not caught by the current unit tests because they don't go through `Responder`.

**3. `anthropic_access_token` stored unencrypted in `Setting`**
The field needs to be added to `ENCRYPTED_FIELDS` alongside the other secret tokens. Storing it in plaintext means DB dumps expose the key.

None of these are blocked by missing VCR cassettes — they're wiring bugs that should be reproducible with existing stubs.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/models/provider/anthropic.rb [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+    client_options[:base_url] = base_url if base_url.present?
+    client_options[:timeout] = ENV.fetch("ANTHROPIC_REQUEST_TIMEOUT", 600).to_i
+
+    @client = ::Anthropic::Client.new(**client_options)
+    @base_url = base_url
+    @default_model = model.presence || DEFAULT_MODEL
+  end
+
+  def supports_model?(model)
+    DEFAULT_ANTHROPIC_MODEL_PREFIXES.any? { |prefix| model.to_s.start_with?(prefix) }
```
[the remark this answers is not in the document]
Comment (gariasf) [reply app/models/provider/anthropic.rb:50]: Fixed in [`714cf0bb`](https://github.com/we-promise/sure/commit/714cf0bb) — `Provider::Anthropic#supports_model?` now bypasses the `claude` prefix gate when `custom_endpoint?` is true, mirroring the OpenAI provider. Bedrock-shaped IDs like `anthropic.claude-sonnet-4-5-20250929-v1:0` route to the Anthropic provider as intended. Test coverage in [`anthropic_test.rb:34`](https://github.com/we-promise/sure/blob/feature/anthropic-foundation/test/models/provider/anthropic_test.rb#L34).

File: app/models/setting.rb [hunk 2]
```diff
@@ -10,6 +10,10 @@ class ValidationError < StandardError; end
   field :openai_uri_base, type: :string, default: ENV["OPENAI_URI_BASE"]
   field :openai_model, type: :string, default: ENV["OPENAI_MODEL"]
   field :openai_json_mode, type: :string, default: ENV["LLM_JSON_MODE"]
+  field :anthropic_access_token, type: :string, default: ENV["ANTHROPIC_ACCESS_TOKEN"].presence || ENV["ANTHROPIC_API_KEY"]
```
[the remark this answers is not in the document]
Comment (gariasf) [reply app/models/setting.rb]: Fixed in [`714cf0bb`](https://github.com/we-promise/sure/commit/714cf0bb) — `:anthropic_access_token` is in `EncryptedSettingFields::ENCRYPTED_FIELDS` alongside `:openai_access_token` and the other provider secrets. The MessageEncryptor wrapper now intercepts reads and writes; existing plaintext values are returned as-is until first rewrite (backward-compatible).

File: test/models/provider/anthropic/message_formatter_test.rb [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  # turn produces an Anthropic-compatible history where tool_use_id pairs back
+  # to the original block.
+  test "ChatParser → ToolCall::Function → MessageFormatter round-trips tool_use_id" do
+    anthropic_response = OpenStruct.new(
+      id: "msg_abc",
+      model: "claude-sonnet-4-6",
+      content: [
+        OpenStruct.new(type: :tool_use, id: "toolu_round_trip", name: "get_net_worth", input: { "currency" => "USD" })
+      ]
+    )
```
[the remark this answers is not in the document]
Comment (gariasf) [reply test/models/provider/anthropic/message_formatter_test.rb:101]: Fixed in [`17e9e0d9`](https://github.com/we-promise/sure/commit/17e9e0d9) — added `require "ostruct"` to the three test files in this PR that reference `OpenStruct`. Same convention as [`hostings_controller_test.rb:2`](https://github.com/we-promise/sure/blob/feature/anthropic-foundation/test/controllers/settings/hostings_controller_test.rb#L2). Equivalent fixes pushed to PR2 ([`8913007c`](https://github.com/we-promise/sure/commit/8913007c)) and PR3 ([`668b0555`](https://github.com/we-promise/sure/commit/668b0555)).

**Comment by gariasf** [comment]
@sure-design Thanks for the second pass. Walking through each — all three are already resolved on the current tip (`17e9e0d9`):

**1. Streaming class namespace mismatch** — this isn't actually a mismatch. The `anthropic` gem ships an explicit alias in [`lib/anthropic/streaming.rb`](https://github.com/anthropics/anthropic-sdk-ruby/blob/main/lib/anthropic/streaming.rb):

```ruby
module Anthropic
  Streaming = Anthropic::Helpers::Streaming
end
```

`::Anthropic::Streaming::TextEvent` and `::Anthropic::Helpers::Streaming::TextEvent` resolve to the **same class object** — Ruby's constant resolution follows the assignment. The `case/when` branches in `stream_chat_response` match correctly via `is_a?` (which is what `Class#===` delegates to).

Empirical confirmation: [`anthropic_test.rb:147`](https://github.com/we-promise/sure/blob/feature/anthropic-foundation/test/models/provider/anthropic_test.rb#L147) creates ad-hoc subclasses of `::Anthropic::Streaming::TextEvent` and `::Anthropic::Streaming::MessageStopEvent`, fires them through the stream proc, and asserts that two text deltas and one final-response chunk are emitted. The test passes — if the namespaces were distinct classes the case/when wouldn't dispatch and the assertions would fail.

**2. `Provider::Anthropic#chat_response` missing `messages:` kwarg** — fixed in [`714cf0bb`](https://github.com/we-promise/sure/commit/714cf0bb). Signature on the current tip ([`anthropic.rb:96`](https://github.com/we-promise/sure/blob/feature/anthropic-foundation/app/models/provider/anthropic.rb#L96)):

```ruby
def chat_response(
  prompt,
  model:,
  instructions: nil,
  functions: [],
  function_results: [],
  messages: nil,           # ← accepted (ignored, Anthropic uses conversation_history)
  conversation_history: [],
  ...
)
```

Regression test at [`anthropic_test.rb:100`](https://github.com/we-promise/sure/blob/feature/anthropic-foundation/test/models/provider/anthropic_test.rb#L100) exercises a Responder-shaped call with both `messages:` and `conversation_history:` kwargs.

**3. `anthropic_access_token` stored unencrypted** — fixed in [`714cf0bb`](https://github.com/we-promise/sure/commit/714cf0bb). `:anthropic_access_token` is in `EncryptedSettingFields::ENCRYPTED_FIELDS` in [`setting.rb:77`](https://github.com/we-promise/sure/blob/feature/anthropic-foundation/app/models/setting.rb#L77) alongside `:openai_access_token`, `:twelve_data_api_key`, `:external_assistant_token`, etc. The MessageEncryptor wrapper at [`setting.rb:80-89`](https://github.com/we-promise/sure/blob/feature/anthropic-foundation/app/models/setting.rb#L80-L89) intercepts reads (decrypt) and writes (encrypt) for every listed field; existing plaintext values fall back via the `InvalidSignature`/`InvalidMessage` rescue and re-encrypt on next write.

Worth re-checking against `17e9e0d9` if the bot was running against an older snapshot.

File: app/models/chat.rb [hunk 4]
```diff
@@ -51,10 +51,15 @@ def generate_title(prompt)
...
+    # Returns the default AI model to use for chats.
+    # Resolved from the configured llm_provider so installs that swap providers
+    # don't have to manually update every chat default.
     def default_model
-      Provider::Openai.effective_model.presence || Setting.openai_model
+      if Setting.llm_provider == "anthropic"
+        Provider::Anthropic.effective_model.presence || Setting.anthropic_model
+      else
+        Provider::Openai.effective_model.presence || Setting.openai_model
+      end
```
[the remark this answers is not in the document]
Comment (gariasf) [reply app/models/chat.rb]: Fixed in [`714cf0bb`](https://github.com/we-promise/sure/commit/714cf0bb). `Chat.default_model` now short-circuits the `prefers_anthropic` branch on `Provider::Anthropic.configured?`, falls through to `elsif Provider::Openai.configured?` next, then `elsif Provider::Anthropic.configured?`, and only lands on the OpenAI default-model fallback when both providers are unconfigured. New `Provider::{Anthropic,Openai}.configured?` class methods centralize the credential check.

Tests at [`chat_test.rb:65-86`](https://github.com/we-promise/sure/blob/feature/anthropic-foundation/test/models/chat_test.rb#L65-L86) cover all three preference + credential combinations and assert against `Provider::*.effective_model` rather than `DEFAULT_MODEL` so they don't flake under `ANTHROPIC_MODEL`/`OPENAI_MODEL` env overrides.

File: app/models/provider/anthropic.rb [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+    instructions: nil,
+    functions: [],
+    function_results: [],
+    conversation_history: [],
+    streamer: nil,
+    previous_response_id: nil,
+    session_id: nil,
+    user_identifier: nil,
+    family: nil
+  )
```
[the remark this answers is not in the document]
Comment (gariasf) [reply app/models/provider/anthropic.rb:110]: Fixed in [`714cf0bb`](https://github.com/we-promise/sure/commit/714cf0bb). Signature now includes `messages: nil` ([`anthropic.rb:96`](https://github.com/we-promise/sure/blob/feature/anthropic-foundation/app/models/provider/anthropic.rb#L96)) — Anthropic ignores it (uses `conversation_history:` instead) but accepts the kwarg so the Responder call stops raising `ArgumentError`. Regression test at [`anthropic_test.rb:100`](https://github.com/we-promise/sure/blob/feature/anthropic-foundation/test/models/provider/anthropic_test.rb#L100) calls `chat_response` with both kwargs to lock this in.

File: app/models/setting.rb [hunk 6]
```diff
@@ -10,6 +10,10 @@ class ValidationError < StandardError; end
...
+  field :anthropic_model, type: :string, default: ENV["ANTHROPIC_MODEL"]
+  field :anthropic_base_url, type: :string, default: ENV["ANTHROPIC_BASE_URL"]
+  field :llm_provider, type: :string, default: ENV.fetch("LLM_PROVIDER", "openai")
```
[the remark this answers is not in the document]
Comment (gariasf) [reply app/models/setting.rb]: Fixed in [`714cf0bb`](https://github.com/we-promise/sure/commit/714cf0bb). `:anthropic_access_token` added to `EncryptedSettingFields::ENCRYPTED_FIELDS` at [`setting.rb:77`](https://github.com/we-promise/sure/blob/feature/anthropic-foundation/app/models/setting.rb#L77) alongside the other provider secrets. MessageEncryptor wrapper at [`setting.rb:80-89`](https://github.com/we-promise/sure/blob/feature/anthropic-foundation/app/models/setting.rb#L80-L89) intercepts reads/writes. Backward-compatible: existing plaintext values fall through the `InvalidSignature`/`InvalidMessage` rescue and re-encrypt on next write.

File: test/models/provider/anthropic_test.rb [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+  test "supports_pdf_processing? true for claude models" do
+    assert @subject.supports_pdf_processing?(model: "claude-sonnet-4-6")
+    assert_not @subject.supports_pdf_processing?(model: "gpt-4o")
+  end
+
+  test "effective_model defers to ENV when set" do
+    ClimateControl.modify("ANTHROPIC_MODEL" => "claude-haiku-4-5") do
+      assert_equal "claude-haiku-4-5", Provider::Anthropic.effective_model
+    end
+  end
```
[the remark this answers is not in the document]
Comment (gariasf) [reply test/models/provider/anthropic_test.rb]: Fixed in [`66753319`](https://github.com/we-promise/sure/commit/66753319). `Provider::Anthropic.effective_model` now uses `ENV["ANTHROPIC_MODEL"].presence || Setting.anthropic_model` ([`anthropic.rb:18`](https://github.com/we-promise/sure/blob/feature/anthropic-foundation/app/models/provider/anthropic.rb#L18)) — the `Setting` lookup only fires when the env var is actually absent.

Test hardened at [`anthropic_test.rb:57-62`](https://github.com/we-promise/sure/blob/feature/anthropic-foundation/test/models/provider/anthropic_test.rb#L57-L62) with `Setting.expects(:anthropic_model).never` so the ENV-precedence assertion is enforced, not just observed.

File: test/models/chat_test.rb [hunk 8]
```diff
@@ -62,6 +62,29 @@ class ChatTest < ActiveSupport::TestCase
...
+    assert_equal Provider::Openai::DEFAULT_MODEL, Chat.default_model
+  end
+
+  test "default_model uses Anthropic when OpenAI is unconfigured" do
+    Provider::Anthropic.stubs(:configured?).returns(true)
+    Provider::Openai.stubs(:configured?).returns(false)
+    Setting.stubs(:llm_provider).returns("openai")
+
+    assert_equal Provider::Anthropic::DEFAULT_MODEL, Chat.default_model
+  end
```
[the remark this answers is not in the document]
Comment (gariasf) [reply test/models/chat_test.rb]: Fixed in [`a0c552cb`](https://github.com/we-promise/sure/commit/a0c552cb). The three `default_model` tests now assert against `Provider::*.effective_model` directly ([`chat_test.rb:65-86`](https://github.com/we-promise/sure/blob/feature/anthropic-foundation/test/models/chat_test.rb#L65-L86)) instead of the `DEFAULT_MODEL` constant — they verify the routing decision (which provider wins) without coupling to the constant value, so `ANTHROPIC_MODEL`/`OPENAI_MODEL` env overrides don't make them flake.

Same brittleness existed on two pre-existing chat tests (`creates with default model when model is nil/empty`) — those now compare against `Chat.default_model` directly. Verified: `ANTHROPIC_MODEL=claude-haiku-4-5 OPENAI_MODEL=gpt-4o bin/rails test test/models/chat_test.rb` → 16 runs / 0 failures (was 2 failures before).

**Comment by gariasf** [comment]
@coderabbitai review again plz

File: app/models/provider/anthropic.rb [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+      langfuse_client.trace(
+        name: name,
+        input: input,
+        session_id: session_id,
+        user_id: user_identifier,
+        environment: Rails.env
+      )
+    rescue => e
+      Rails.logger.warn("Langfuse trace creation failed: #{e.message}\n#{e.full_message}")
+      nil
```
[the remark this answers is not in the document]
Comment (gariasf) [reply app/models/provider/anthropic.rb]: Fixed in [`5ab9bb33`](https://github.com/we-promise/sure/commit/5ab9bb33). All three Langfuse rescue branches in `Provider::Anthropic` now log `#{e.class}: #{e.message}` only, dropping `e.full_message` which bundles the backtrace + cause chain and on some SDK error types includes serialized request/response payloads (user prompt, model output). Sites updated: `create_langfuse_trace`, `log_langfuse_generation`, `upsert_langfuse_trace`.

Note: `Provider::Openai` has the same pattern (copy-pasted source). Harmonization deferred to a follow-up cleanup PR to keep this commit scoped.

File: app/models/provider/anthropic/message_formatter.rb [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    def parse_arguments(arguments)
+      case arguments
+      when nil then {}
+      when Hash then arguments
+      when String
+        return {} if arguments.blank?
+        JSON.parse(arguments)
+      else arguments
+      end
+    rescue JSON::ParserError
```
[the remark this answers is not in the document]
Comment (gariasf) [reply app/models/provider/anthropic/message_formatter.rb:129]: Fixed in [`5ab9bb33`](https://github.com/we-promise/sure/commit/5ab9bb33). `MessageFormatter#parse_arguments` now coerces any non-Hash parsed result to `{}` — scalars, booleans, arrays, and non-String non-Hash inputs all yield an empty object so `tool_use.input` is always a JSON map as Anthropic requires.

Normal flow stores Hash arguments end-to-end (input came from Anthropic as a Hash, persisted via `function_request.function_args = input.to_json`, parsed back with `JSON.parse`), so this is defensive against corrupt/legacy rows or cross-provider ToolCall bleed. Two new test cases at [`message_formatter_test.rb:175-211`](https://github.com/we-promise/sure/blob/feature/anthropic-foundation/test/models/provider/anthropic/message_formatter_test.rb#L175-L211) cover scalar/bool/array JSON strings and non-String non-Hash arguments.

**Comment by gariasf** [comment]
Latest review round addressed in [`5ab9bb33`](https://github.com/we-promise/sure/commit/5ab9bb33):

- 🟠 Sanitized Langfuse warn logs (sensitive payload leak) — inline reply on the affected thread.
- 🟠 Normalized `parse_arguments` so `tool_use.input` is always a Hash — inline reply on the affected thread.
- 🟡 Deduped the chat-history fetch in `Assistant::Responder`. New memoized `complete_chat_messages` runs once per LLM turn; `chat_message_records` filters out the current message via `Array#reject`, `openai_messages_payload` iterates the cached array. One SQL query per turn instead of two. (This was in the review-body Nitpick section without a dedicated inline thread, hence the summary here.)

Stack rebased onto the latest `c120a050` (main merge), all 4409 tests green, rubocop + brakeman clean.

**Comment by sure-design** [comment]
The `security-scan` CI check is failing on the latest commit. All other CI checks pass. Please investigate and resolve the security scan failure before this can merge.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by jjmata** [comment]
One reason that the LLM forgot to mention why we need Anthopic models [is this](https://petergpt.github.io/bullshit-benchmark/viewer/index.v2.html):
Screenshot 2026-05-31 at 4 09 01 PM

**Review by jjmata — APPROVED** [review]
